src/tools/genpolicy/rules.rego

```diff
@@ -121,6 +128,47 @@ allow_create_container_input {
     print("allow_create_container_input: true")
 }
 
+# value hasn't been seen before, save it to state
+state_allows(key, value) = action {
+  state := get_state()
+  not state[key]
+  print("state_allows: saving to state key =", key, "value =", value)
+  path := get_state_path(key) 
+  action := {
+    "op": "add",
+    "path": path, 
+    "value": value,
+  }
+}
+
+# value matches what's in state, allow it
+state_allows(key, value) = action {
+  state := get_state()
+  value == state[key]
+  print("state_allows: found key =", key, "value =", value, " in state")
+  action := null
+}
+
+# helper functions to interact with the state
+get_state() = state {
+  state := data["pstate"]
+}
+
+get_state_path(key) = path {
+    path := sprintf("/pstate/%s", [key]) # path := "/" + key
```

Comment:
Hackers traditionally liked to attack sprintf (for C apps). Let's use concat() instead - as other rules do to.

Reply:
Updated back to use concat